Nemo API

arporter opened the pull request:
Initial support for the NEMO API.

arporter commented:
Ready for first review. This PR is by no means "complete" but it's already big enough. I haven't added much in the way of documentation and the functionality is incomplete. However, there is full test coverage and I've added two new examples.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -1914,6 +1928,49 @@ def _encloses_omp_directive(self):
+        # Find the locations in which we must insert the begin/end
+        # directives...
+        # Find the children of this node in the AST of our parent node
+        try:
```
I'm sure you don't need me to tell you that this insertion is a bit messy and we need to better support this sort of thing in fparser2, especially adding directives.

arporter replied:
Agreed.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -1914,6 +1928,49 @@ def _encloses_omp_directive(self):
+        # the AST representing our last child
+        enddir = Comment(FortranStringReader("!$omp end parallel",
+                                             ignore_comments=False))
+        if end_idx == len(self._parent._ast.content) - 1:
```
It's unclear to me why you have this if clause here

arporter replied:
Because I assumed that `some_list.insert(len(some_list), my_obj)` would be an error but it isn't - I've removed the if clause.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -2081,6 +2137,67 @@ def gen_code(self, parent):
+                "but this Node has {0} children: {1}".
+                format(len(self._children), self._children))
+
+        # Find the locations in which we must insert the begin/end
```
Good comments. Thanks.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -1264,6 +1271,13 @@ def is_openmp_parallel(self):
     def gen_code(self):
         raise NotImplementedError("Please implement me")
 
+    def update(self):
```
You add support for OMP PARALLEL as well as OMP PARALLEL DO but you don't use the former in an example. I've not checked the tests yet.

arporter replied:
I've just checked and I do have tests for it.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -2325,8 +2442,10 @@ def loop_type(self):
     def loop_type(self, value):
-        assert value in self._valid_loop_types, \
-            "Error, loop_type value is invalid"
+        if value not in self._valid_loop_types:
```
raise generation error rather than an assert. Good.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -3555,3 +3674,54 @@ def name(self):
         return None, None
+
+
+class IfBlock(Node):
```
Added if and elseif as nodes. Good.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+        None otherwise. '''
+        kernels = self.walk(self.children, NemoKern)
+        if kernels:
+            # TODO cope with case where loop contains >1 kernel (e.g.
```
We should raise an exception here, not just blindly continue!

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+    def __init__(self, ast, parent=None):
+        from fparser.common.readfortran import FortranStringReader
+        Loop.__init__(self, parent=parent,
+                      valid_loop_types=VALID_LOOP_TYPES+["unknown"])
```
perhaps unknown should be added to the main list instead of adding it as a string in more than one place (which is dangerous).

arporter replied:
Good point. Done.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+                # and thus we need an explicit loop for this dimension.
+                outermost_dim = idx
+
+        # TODO allow for implicit loops with specified bounds (e.g. 2:jpjm1)
```
If this is not supported then we should raise an exception if this situation arises.

arporter replied:
Done + test.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+        self._step = 1
+        if outermost_dim == 0:
+            # TODO ensure no name clash is possible with variables that
+            # already exist in the NEMO source.
```
Good point. Please add to the list of issues.

arporter replied:
Done.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+        # TODO allow for implicit loops with specified bounds (e.g. 2:jpjm1)
+        self._start = 1
+        self._step = 1
+        if outermost_dim == 0:
```
We should parameterise the ordering to allow for different ones i.e. have a lookup based on a config file.

arporter replied:
Good point. I've added it as a dict at the top of the module for now. As mentioned earlier, I'm unsure about whether this belongs in a config file.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+                                format(names[0]))
+        # TODO check that this variable has not already been declared
+        # Requires that we capture all variable declarations in the routine
+        # in some sort of management class.
```
the namespace manager can do this if I understand correctly.

arporter replied:
Oh yes, that's an idea. I was thinking I might want to store more information than just the name (e.g. type, rank etc) but I'm not sure yet...

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+            subsec.items = tuple(indices)
+
+        # Create an explicit loop
+        text = '''\
```
It's a small thing but I prefer using
(
"xxx\n"
"yyy\n"
)
as it allows us to inline the code more nicely.

arporter replied:
Done.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+do {0}=1,{1},{2}
+  replace = me
+end do
+'''.format(self._variable_name, self._stop, self._step)
```
Actually we need a higher level (f2pytrans style) approach to create a loop, or extend the fparser API. But that is clearly future work.

arporter replied:
Absolutely.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+        #    array(:,:,jk) = some_expression
+        # so we check the left-hand side...
+        lhs = node.items[0]
+        if isinstance(lhs, Fortran2003.Part_Ref):
```
Again I think that whether a statement uses array syntax or not should be provided by a library.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+                                       parent=self))
+
+    def gen_code(self):
+        ''' Override abstract method of base class. '''
```
raise exception if this is called?

arporter replied:
Done + test.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+                           nodes_parent=self._ast._parent)
+
+    def gen_code(self):
+        ''' Override abstract method of base class. '''
```
raise exception?

arporter replied:
Done + test.

arporter commented on doc/nemo.rst:
```diff
@@ -0,0 +1,96 @@
+In contrast to the other APIs supported by PSyclone, the NEMO API is
+designed to work with source code that does *not* follow the PSyKAl
+separation of concerns. Instead, the NEMO source code is treated as if
+it were a manually written PSy layer with all kernels in-lined. This
```
Not really. In PSyKAl the Algorithm layer contains calls to `invoke()` that are then turned into calls to generated PSy layer routines. In NEMO we are treating each subroutine as an `invoke` that consists of zero or more (inlined) kernel calls. As such we never consider the algorithm.

arporter commented on doc/nemo.rst:
```diff
@@ -0,0 +1,96 @@
+         Loop[type='levels',field_space='None',it_space='None']
+             Loop[type='lat',field_space='None',it_space='None']
+                 Loop[type='lon',field_space='None',it_space='None']
+                     KernCall[]
```
Created #222.

arporter commented on doc/nemo.rst:
```diff
@@ -0,0 +1,96 @@
+Algorithm
+---------
+
+Since NEMO source is treated as a pre-existing PSy layer, this API
```
Halo exchanges only exist in in the PSy layer so I think that's correct. Control code is more difficult but I don't see how we could create an Algorithm layer?

arporter commented on examples/nemo/eg1/runme_openmp.py:
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python
+## -------------------------------------------------------------------------
```
Oops! Done.

arporter commented on examples/nemo/eg1/runme_openmp.py:
```diff
@@ -0,0 +1,53 @@
+
+    TRANS_INFO = TransInfo()
+    print(TRANS_INFO.list)
+    FUSE_TRANS = TRANS_INFO.get_trans_name('LoopFuse')
```
Oops. Removed.

arporter commented on examples/nemo/eg1/runme_openmp.py:
```diff
@@ -0,0 +1,53 @@
+
+    for loop in sched.loops():
+        kernel = loop.kernel
+        if kernel:
```
Good point.

arporter commented on examples/nemo/eg1/runme_openmp.py:
```diff
@@ -0,0 +1,53 @@
+        kernel = loop.kernel
+        if kernel:
+            if loop.loop_type == "levels":
+                sched, _ = OMP_TRANS.apply(loop)
```
I've extended the documentation and added READMEs.

arporter commented on examples/nemo/eg1/runme_openmp.py:
```diff
@@ -0,0 +1,53 @@
+    sched.view()
+
+    psy.invokes.get('tra_adv').schedule = sched
+    print(psy.gen)
```
There's an x-failing test for the first problem. I've created a new `Limitations` section in the NEMO API documentation.

arporter commented on examples/nemo/eg2/omp_levels_trans.py:
```diff
@@ -0,0 +1,60 @@
+    for loop in sched.loops():
+        if loop.kernel and loop.loop_type == "levels":
+            sched, _ = ompt.apply(loop)
+    psy.invokes.get('tra_ldf_iso').schedule = sched
```
Oops, yes. (Although, since we modify the psy we're passed a reference to, it isn't strictly necessary.)

arporter commented on examples/nemo/eg2/runme_openmp.py:
```diff
@@ -0,0 +1,80 @@
+
+    TRANS_INFO = TransInfo()
+    print(TRANS_INFO.list)
+    FUSE_TRANS = TRANS_INFO.get_trans_name('LoopFuse')
```
Oops. Tidied.

arporter commented on examples/nemo/eg2/runme_openmp.py:
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python
```
I've added the CASE issue to the list of limitations. The `else` *is* a child of the `if` since it cannot exist in isolation.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -979,6 +979,34 @@ def apply(self, node):
         return OMPParallelLoopTrans.apply(self, node)
 
 
+class NemoOMPParallelLoopTrans(OMPParallelLoopTrans):
```
Removed.

arporter commented on src/psyclone/parse.py:
```diff
@@ -906,6 +906,12 @@ def parse(alg_filename, api="", invoke_name="invoke", inf_name="inf",
+        # For this API we just parse the NEMO code and return the resulting
+        # fparser2 AST
+        ast = parse_nemo(alg_filename)
+        return ast, ast
```
Good point. I've changed it to return None for the algorithm AST.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+NEMO_SCHEDULE_COLOUR_MAP["CodeBlock"] = "red"
+
+# The loop index variables we expect NEMO to use
+NEMO_LOOP_VARS = ["ji", "jj", "jk", "jt", "jn"]
```
My view of config files has changed a bit after various discussions with Joerg. Although I agree we want to code for future flexibility and adaptability, I'm not sure that moving these into a config file would significantly improve either of those. Do we expect the average user to want to modify these quantities or is it more of a PSyclone developer issue?

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+    Mixin class to provide functionality for processing the fparser2 AST.
+    '''
+    @staticmethod
+    def add_code_block(parent, statements):
```
I've changed it to `nodes_to_code_block()`

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+    def add_code_block(parent, statements):
+        '''
+        Create a NemoCodeBlock for the supplied list of statements
+        and then wipe the list of statements.
```
Done.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+    # parent information (fparser/#102).
+    def process_nodes(self, parent, nodes, nodes_parent):
+        '''
+        Create the PSyclone AST representation of the code represented
```
Yes. I've added documentation for the interface to this routine too.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+
+    # TODO remove nodes_parent argument once fparser2 AST contains
+    # parent information (fparser/#102).
+    def process_nodes(self, parent, nodes, nodes_parent):
```
Thanks - the benefit of implementing this several times is that I finally got to something good :-)

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+                self.add_code_block(parent, code_block_nodes)
+                parent.addchild(NemoIfBlock(child, parent=parent))
+            elif not isinstance(child, (Fortran2003.End_Do_Stmt,
+                                        Fortran2003.Nonlabel_Do_Stmt)):
```
What about it? :-) I've added this to the list of limitations.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+    def __init__(self, ast, name):
+        self._schedule = None
+        self._name = name
+        self._psy_unique_vars = ["a_variable"]
```
I don't know! Removed.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+
+
+class NemoPSy(PSy):
+    ''' The NEMO 0.1-specific PSy class. This creates a NEMO-specific
```
Oh yes, and interface to constructor documented.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+        Node.__init__(self, parent=parent)
+        # Store a list of the parser objects holding the code
+        # associated with this block
+        self._statements = statements[:]
```
The contents of the list need to be copied because the list itself is temporary and is deleted (in the renamed `add_code_block()` routine). Possibly this indicates something sub-optimal in the way I've implemented code block generation but I can't quite see what it is.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+    def _load_from_implicit_loop(self, loop, parent=None):
+        ''' Populate the state of this NemoKern object from an fparser2
+        AST for an implicit loop (Fortran array syntax) '''
+        # TODO implement this method!
```
Added to the list of issues in the doc.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,822 @@
+        (and must therefore be e.g. threadprivate if doing OpenMP)
+
+        '''
+        return []
```
Added to the list in the doc.

arporter commented:
OK, I think I've addressed all comments. I've also gone through and made sure that all the docstrings are up-to-date in the new module.
Ready for 2nd review...

arporter commented on examples/nemo/eg2/README:
```diff
@@ -0,0 +1,52 @@
+
+    $ psyclone -api "nemo" -s ./omp_levels_trans.py traldf_iso.F90
+
+Again, the generated Fortran will be written to stdout.
```
Done.

arporter commented on .travis.yml:
```diff
@@ -45,12 +45,16 @@ env:
   - ./bin/install_optional.sh fparser_submodule
   - pip install coveralls
   - pip install codecov
+# TODO remove following two lines (which install the master branch of
```
Added #235 and put this in the comment above and in .travis.yml

arporter commented on examples/nemo/eg2/omp_levels_trans.py:
```diff
@@ -0,0 +1,60 @@
+
+Once you have PSyclone installed, this script may be used by doing:
+
+ >>> psyclone -api "nemo0.1" -s ./omp_levels_trans.py traldf_iso.F90
```
Done.

arporter commented:
* If we do not monkeypatch the Assignment_Stmt in parse_test then we are fine
* If we monkeypatch but do not actually run `get_integer_array()` then we are fine
* If I put a print/pdb inside the patched function it is only called from within parse_test, not subsequently (which is right because the function I've used to patch goes out of scope)

arporter commented:
* Explicitly calling the `undo()` method of the monkeypatch object makes no difference.
Created a new test file containing the key test from parser_test and the first failing nemo test.
If I override the base-class `__new__` method and put in a pdb call there, I can see that it gets called if I run the 2nd (failing) test on its own. If I run both tests then it does not get called in the second test.

rupertford commented:
Travis is happy.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,954 @@
+                break
+            self._body.append(content)
+
+        #  I could get this by walking back up the tree and counting how
```
So presumably we want to know if we have loop nesting of 2 or 3? Where would this be used and is this part of the todo list?

arporter replied:
I'm not sure of the relevance of this anymore - the way we analyse the code is more sophisticated than it was when I first wrote this (I think). The kernel type is in fact currently only used in the `view` method of the kernel. I've changed this to set `_kernel_type = "Explicit"` for the time being.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,954 @@
+        self._first_private_vars = set()
+        self._private_vars = set()
+        #  If there are scalar variables that are inputs to the DAG (other than
+        #  the loop counters) then they must be declared first-private.
```
Please add "in an OpenMP loop directive."

arporter replied:
Done.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,954 @@
+        # TODO bring habakkuk up-to-date with changes to fparser and then
+        # uncomment this code.
+
+        #  Analyse the loop body to identify private and shared variables
```
for use by OpenMP?

arporter replied:
Added.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -253,6 +254,11 @@ def create(self, invoke_info):
         elif self._type == "gocean1.0":
             from psyclone.gocean1p0 import GOPSy
             return GOPSy(invoke_info)
+        elif self._type == "nemo":
```
Please add argumet information to the docstring in the parent method.

arporter replied:
Done.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -1058,6 +1064,7 @@ def __init__(self, children=None, parent=None):
         else:
             self._children = children
         self._parent = parent
+        self._ast = None  # Reference into fparser2 AST (if any)
```
Please add docstring argument information for the method (either in the __init__ or the class docstring).

arporter replied:
Done. I've also moved the constructor (and `__str__`) up to the start of the class definition as that feels more logical.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -2321,8 +2436,10 @@ def loop_type(self):
     def loop_type(self, value):
-        assert value in self._valid_loop_types, \
-            "Error, loop_type value is invalid"
+        if value not in self._valid_loop_types:
```
Docstring argument info required.

arporter replied:
Oh yes. Done.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -3697,3 +3814,54 @@ def name(self):
+class IfBlock(Node):
+    '''
+    Class representing an if-block within a Schedule.
+    '''
```
Need to describe parent argument in a docstring.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -3697,3 +3814,54 @@ def name(self):
+
+        :return: the name of this node type, possibly with control codes
+                 for colour
+        :rtype: string
```
Isn't it str? And could you include this as part of the return seeing as it' s a simple type?

arporter replied:
It is (str) and Stackoverflow says I can't :-( (http://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists). Updated.

rupertford replied:
Oops, rtfm. Although it would be nicer.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -3697,3 +3814,54 @@ def name(self):
+        return colored("If", SCHEDULE_COLOUR_MAP["If"])
+
+    def view(self, indent=0):
+        ''' Print representation of this node to stdout '''
```
document indent arg.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -3697,3 +3814,54 @@ def name(self):
+class IfClause(IfBlock):
+    '''
+    Represents a sub-clause of an If block - e.g. an "else if()".
+    '''
```
document parent argument.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -3697,3 +3814,54 @@ def name(self):
+
+        :return: the name of this node type, possibly with control codes
+                 for colour
+        :rtype: string
```
str not string? and include as part of return?

arporter replied:
As before: yes and no.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -0,0 +1,954 @@
+        Generate the (updated) fparser2 AST for the NEMO code represented
+        by this NemoPSy object.
+
+        :rtype: :py:class:`fparser.two.Fortran2003.Main_Program` or \
```
Oops, so I do.

arporter commented:
The documentation problem was because of a bug in sphinx. As a workaround I've renamed the file containing the bibitems to be "zz_references.bib" so that sphinx processes it *after* all of the other files.

arporter commented:
In going through and checking pylint I've realised that in this issue I've added an `_ast` property to the `Node` class, without an associated getter. However, due to the kernel-transformation support, the `psyGen.Kern` class already has an `_fp2_ast` property and an `ast` getter. Fixing that here is going to prolong the agony of this Issue so I've created #241 and added this to the list of know issues/restrictions in nemo.rst.

arporter commented:
Travis is happy. Ready for another review...

arporter commented:
I'd misunderstood the sphinxcontrib-bibtex workaround - I should have made sure that the rst file containing the `.. bibliography` command was processed last, not the file containing the bib entries.
Getting one (and only one) "Bibliography" section heading in both latex and html output was also tricky as latex makes such a heading automatically but html does not. I've got around that by using a `.. only:: html` to protect the heading entry in the rst file.
